### ISSUE TITLE
Logging transitions with DRY code?

### ISSUE THREAD

**Issue by awendt** [aasm/aasm#221]
With a fairly complex state machine, we'd like to log transitions so we know how we ended up in the current user state. For each transition, we'll need:

- `aasm.from_state`
- `aasm.to_state`
- `aasm.current_event`

According to the docs:

> During the transition's `:after` callback (and reliably only then) you can access the originating state (the from-state) and the target state (the to state)

This means we'll go from this:

      transitions from:   :initialized,
                  to:     :mismatched

to this:

      transitions from:   :initialized,
                  to:     :mismatched,
                  after:  Proc.new { |*args|
                    track(
                      'from_state'         => aasm.from_state,
                      'into_state'         => aasm.to_state,
                      'statemachine_event' => aasm.current_event
                    )
                  }

**for each transition** which is undesirable from a maintenance perspective. I'm aware of things like `paper_trail` for auditing AR but that does not fit into our current architecture.

So I'm wondering if there's a way to get this done while staying DRY? (I'm leaning toward `alias_method_chain` to add the callback—but there may be another, AASM-supported way I'm not aware of.)

**Comment by alto** [comment]
@awendt There's already a ticket for something similar (take a look at #108). Would that help you?

What you could do now already is to use methods instead of a `Proc`, like this:

```ruby
aasm do
  transitions from: 	:initialized,
              to:   	:mismatched,
              after:	:tracking
end

def	tracking(*args)
  track(
    'from_state'         => aasm.from_state,
    'into_state'         => aasm.to_state,
    'statemachine_event' => aasm.current_event
  )
end
```

**Comment by awendt** [comment]
@alto Thanks for your quick reply. Anything that involves me declaring "I want this code to run after every transition" just _once_ helps here :)

I found #108 but failed to see a connection. After more careful reading I found [this comment](https://github.com/aasm/aasm/issues/108#issuecomment-69034662):

> FYI the reason I raised the original issue was to implement an observer pattern on transitions. (Of course possible to observe the state variable but not the transitions).

So, yes, that exactly describes the issue I'm having.

I think I will switch to symbols and add `alias_method_chain` for the time being but it would be great to have AASM support that in the long run.

**Comment by waynn** [comment]
@awendt How exactly are you using alias_method_chain with this? I'm running into the same problem, are you including it in the `aasm` declaration?

**Comment by awendt** [comment]
@waynn I settled for a duplicated `after:  :track_event` for each transition. Sorry.

**Comment by alto** [comment]
Please take a look at #270 and let me know if that works for your problem. Will release soon.

**Comment by waynn** [comment]
That looks good!

**Comment by awendt** [comment]
@alto 

:+1: — should I close this then?

**Comment by alto** [comment]
Closed with #270 .

**Comment by alto** [comment]
I just released version `4.4.0` including the change.
